### PR TITLE
New Feature: Table Diffing

### PR DESCRIPTION
**Pull request by jschroed91**
New Feature: Table Diffing

There is still quite a bit of work required on this before I would consider it "production-ready" for an open source release, but I'm hoping it will do for the project we're using this library on and will be a decent start on improving how tables are diffed in this library.

### CODE REVIEW

File: lib/Caxy/HtmlDiff/HtmlDiff.php [hunk 1]
```diff
@@ -15,8 +19,20 @@ class HtmlDiff extends AbstractDiff
...
         'table' => '[[REPLACE_TABLE]]',
+        'strong' => '[[REPLACE_STRONG]]',
+        'b' => '[[REPLACE_B]]',
+        'em' => '[[REPLACE_EM]]',
+        'i' => '[[REPLACE_I]]',
         'a' => '[[REPLACE_A]]',
     );
+    protected $useTableDiffing = true;
+
+    public function setUseTableDiffing($bool)
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/HtmlDiff.php]: I don't mean to be a stickler, but annotations.

File: lib/Caxy/HtmlDiff/HtmlDiff.php [hunk 2]
```diff
@@ -243,6 +261,14 @@ protected function diffList($oldText, $newText)
         return $diff->build();
     }
 
+    protected function diffTables($oldText, $newText)
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/HtmlDiff.php]: Annotations

File: lib/Caxy/HtmlDiff/HtmlDiff.php [hunk 3]
```diff
@@ -338,6 +364,13 @@ protected function isPlaceholderType($text, $types, $strict = true)
         return in_array($text, $criteria, $strict);
     }
 
+    protected function isTablePlaceholder($text)
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/HtmlDiff.php]: Annotations

File: lib/Caxy/HtmlDiff/Table/AbstractTableElement.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+namespace Caxy\HtmlDiff\Table;
+
+abstract class AbstractTableElement
+{
+    /**
+     * @var \DOMElement
+     */
+    protected $domNode;
+
+    public function __construct(\DOMElement $domNode = null)
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/AbstractTableElement.php]: Annotations

File: lib/Caxy/HtmlDiff/Table/AbstractTableElement.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $this->domNode = $domNode;
+    }
+
+    public function getDomNode()
+    {
+        return $this->domNode;
+    }
+
+    public function setDomNode(\DOMElement $domNode)
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/AbstractTableElement.php]: Annotations

File: lib/Caxy/HtmlDiff/Table/AbstractTableElement.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     * @var \DOMElement
+     */
+    protected $domNode;
+
+    public function __construct(\DOMElement $domNode = null)
+    {
+        $this->domNode = $domNode;
+    }
+
+    public function getDomNode()
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/AbstractTableElement.php]: Annotations @return

File: lib/Caxy/HtmlDiff/Table/DiffRowPosition.php [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        if ($this->isColumnLessThanOther('new')) {
+            return 'new';
+        } elseif ($this->isColumnLessThanOther('old')) {
+            return 'old';
+        }
+
+        return null;
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/DiffRowPosition.php]: Probably should go through and add all the annotations needed for functions.

File: lib/Caxy/HtmlDiff/Table/Table.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        }
+
+        if ($newRow >= 0 && $newCell >= 0) {
+            return new TablePosition($newRow, $newCell);
+        }
+
+        return null;
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/Table.php]: Same about annotations in this file.

File: lib/Caxy/HtmlDiff/Table/TableCell.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    public function getColspan()
+    {
+        return $this->getAttribute('colspan') ?: 1;
+    }
+
+    public function getRowspan()
+    {
+        return $this->getAttribute('rowspan') ?: 1;
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/TableCell.php]: Annotations for all functions.

File: lib/Caxy/HtmlDiff/Table/TableDiff.php [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+                    $thresholdCount += $increment;
+                }
+            }
+        }
+
+        $matchPercentage = ($totalCount > 0) ? ($thresholdCount / $totalCount) : 0;
+
+        return $matchPercentage;
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/TableDiff.php]: Annotations

File: lib/Caxy/HtmlDiff/Table/TableMatch.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    public function getEndInOld()
+    {
+        return $this->endInOld;
+    }
+
+    public function getEndInNew()
+    {
+        return $this->endInNew;
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/TableMatch.php]: Annotations. Sorry!

File: lib/Caxy/HtmlDiff/Table/TablePosition.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static function compare($a, $b)
+    {
+        if ($a->getRow() == $b->getRow()) {
+            return $a->getCell() - $b->getCell();
+        }
+
+        return $a->getRow() - $b->getRow();
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/TablePosition.php]: Annotations

File: lib/Caxy/HtmlDiff/Table/TableRow.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    public function insertCells($cells, $position = null)
+    {
+        if ($position === null) {
+            $this->cells = array_merge($this->cells, $cells);
+        } else {
+            array_splice($this->cells, $position, 0, $cells);
+        }
+    }
+}
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/TableRow.php]: Annotations file-wide.

File: lib/Caxy/HtmlDiff/Table/Table.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public function insertRows($rows, $position = null)
+    {
+        if ($position === null) {
+            $this->rows = array_merge($this->rows, $rows);
+        } else {
+            array_splice($this->rows, $position, 0, $rows);
+        }
+    }
```
Comment (adamCaxy) [inline lib/Caxy/HtmlDiff/Table/Table.php]: Is this supposed to return something?

**Comment by adamCaxy** [comment]
I can't really see anything else beyond annotations.

**Comment by jschroed91** [comment]
@adamCaxy Review addressed, if you want to take another look.

**Comment by adamCaxy** [comment]
:pineapple: Looks good to me. Only thing was the annotations.

**Pull request merged**